packages/react/src/useFetch.ts

```diff
@@ -0,0 +1,146 @@
+ *
+ * @example
+ * ```
+ * export function UserByEmail(props: { email: string }) {
+ *   const [{data, fetching, error}, refresh] = useFetch("/users/get", {
+ *     method: "GET",
+ *     body: JSON.stringify({ email: props.email }})
+ *     headers: {
+ *       "content-type": "application/json",
+ *     }
+ *     json: true,
+ *   });
+ *
+ *   if (result.error) return <>Error: {result.error.toString()}</>;
+ *   if (result.fetching && !result.data) return <>Fetching...</>;
+ *   if (!result.data) return <>No user found with id={props.id}</>;
+ *
+ *   return <div>{result.data.name}</div>;
+ * }
+ */
+export function useFetch(path: string, options: { stream: true } & FetchHookOptions): FetchHookResult<ReadableStream<Uint8Array>>;
+export function useFetch<T extends Record<string, any>>(url: string, options: { json: true } & FetchHookOptions): FetchHookResult<T>;
+export function useFetch(path: string, options?: FetchHookOptions): FetchHookResult<string>;
+export function useFetch<T = string>(path: string, options?: FetchHookOptions): FetchHookResult<T> {
+  // Used to prevent state update if the component is unmounted
+  const mounted = useRef<boolean>(true);
+  const memoizedOptions = useStructuralMemo<FetchHookOptions>(options ?? {});
+  const connection = useConnection();
+
+  const [state, dispatch] = useReducer<Reducer<FetchHookState<T>, FetchAction<T>>, FetchHookOptions>(
+    reducer,
+    memoizedOptions,
+    (memoizedOptions) => {
+      return { fetching: startRequestByDefault(memoizedOptions), options: memoizedOptions };
+    }
+  );
+
+  const send = useCallback(
+    async (sendOptions?: Partial<FetchHookOptions>): Promise<T> => {
+      dispatch({ type: "fetching" });
+
+      let data: any;
+      let response: Response | undefined = undefined;
+
+      try {
+        // make the fetch call using GadgetConnection to pass along auth and other headers
+        response = await connection.fetch(path, { ...memoizedOptions, ...sendOptions });
+        if (!response.ok) {
+          throw new Error(response.statusText);
+        }
+
+        if (memoizedOptions?.json) {
+          data = await response.json();
+        } else if (memoizedOptions?.stream) {
+          data = response.body;
+        } else {
+          data = await response.text();
+        }
+
+        if (!mounted.current) return data;
+
+        dispatch({ type: "fetched", payload: data });
+      } catch (error: any) {
+        if (!mounted.current) return null as any;
+
+        const wrapped = ErrorWrapper.forClientSideError(error, response);
+        dispatch({ type: "error", payload: wrapped });
+        throw wrapped;
+      }
+      return data;
+    },
+    [memoizedOptions, path]
+  );
+
+  useEffect(() => {
+    mounted.current = true;
+    if (startRequestByDefault(memoizedOptions)) {
+      void send().catch(() => {
+        // report error via return state
+      });
+    }
+
+    return () => {
+      mounted.current = false;
```

Comment:
Does the fetcher support an `AbortSignal` param? Would it be worth adding one?

Reply:
It does for sure! End users can pass in their own abort signal if they want to. I figured it was a lot of complexity for us to support both them passing a signal and us managing a signal internally, but I can do it if it seems super necessary. Was hoping to get something out but I think in the new AI-y world aborting streaming responses makes a lot of sense.

Reply:
I'm game for holding off and seeing how it goes. I'm gonna guess the typical use case would be to abort on unmount, so if we see that happening or hear our users mention that, we can bake it in 🚀